totemsrp: Fix warnings produced by gcc 9.1

jfriesse opened the pull request:
New gcc warn about passing posibly unaligned pointer from packed
structure. This shouldn't be problem for x86.

Implemented solution is to let compiler do its job (compiler knows if
pointer is aligned so accessing structure field is safe) and
add macro alternative of srp_addr_copy called
srp_addr_copy_aligned_macro where required, or use temporary (compiler
aligned) variable where unaligned content is copied.

Signed-off-by: Jan Friesse <jfriesse@redhat.com>

jfriesse commented:
I'm not entirely happy with this patch and if somebody has a better solution (without using macros) then I would be really happy.

Otherwise eventho the patch is quite long it is quite obvious what is happening but it's in super sensitive part of the code so second pair of eyes really needed.

chrissie-c commented:
One solution to using a macro (I'm not keen on those either) would be to pass the srp_addr by value rather than reference. As it's only a nodeid now it won't put pressure on the stack - even less so if you declare it 'inline'

jfriesse commented:
@chrissie-c Yep, I was thinking about it too and it would be usable just fine if unaligned srp_addr would be used as a _source_ of srp_addr_call. Sadly it is not usable in this context because member of packed structure is _destination_ of srp_copy_addr call so it needs to be pointer.

chrissie-c commented:
D'oh! yes, of course. Sorry.

jfriesse commented:
Pushed updated version of patch. Thank you @chrissie-c for idea of using function which returns structure (not used in srp_addr_copy but now used for new function srp_addr_endian_convert) which lead me to use simple assignment. I (personally) like this solution a LOT more than previous one.

Changed commit msg:

New gcc warn about passing posibly unaligned pointer from packed
structure. This shouldn't be problem for x86.

Implemented solution is to let compiler do its job (compiler knows if
pointer is aligned so accessing structure field is safe) and
use it together with support for asigning and returning of structure
(not a pointer to the structure).

- srp_addr_copy is removed and replaced by simple assignment
- srp_addr_copy_endian_convert is removed and replaced by
  srp_addr_endian_convert function which takes srp_addr structure and
  returns endian converted srp_addr structure
- functions which accepts srp_addr array are not changed because
  (luckily) non-aligned pointer is always just one item array and
  such item is always used as a source pointer so it's possible to use
  temporary variable

jfriesse commented:
@chrissie-c Thank you for the review and base of idea of let the compiler do its job and use simple structure assignment.

Merged now as a b0c24ec6652f91f6135c32d3bc1524cac5817102
